### PR TITLE
zap: implement wrapper

### PR DESCRIPTION
**Pull request by karasz**
Signed-off-by: Nagy Károly Gábriel <k@jpi.io>

### CODE REVIEW

File: zap/zap.go [hunk 1]
```diff
@@ -14,89 +15,173 @@ var (
...
 }
 
 // Enabled tells this logger is enabled
-func (zl *Logger) Enabled() bool {
+func (zpl *Logger) Enabled() bool {
+	if zpl == nil || zpl.logger == nil || zapcore.LevelOf(zpl.logger.Core()) == zapcore.InvalidLevel {
+		return false
+	}
+
 	return true
```
Comment (amery) [inline zap/zap.go:30]: shouldn't we be calling `zpl.logger.Check()` here to confirm if a log entry of this level would be logged?

File: zap/zap.go [hunk 2]
```diff
@@ -14,89 +15,173 @@ var (
...
-	zl.print(fmt.Sprintln(args...))
+func (zpl *Logger) Println(args ...any) {
+	zpl.print(fmt.Sprintln(args...))
 }
 
 // Printf adds a log entry with arguments handled in the manner of fmt.Printf
-func (zl *Logger) Printf(format string, args ...any) {
-	zl.print(fmt.Sprintf(format, args...))
+func (zpl *Logger) Printf(format string, args ...any) {
+	zpl.print(fmt.Sprintf(format, args...))
```
Comment (amery) [inline zap/zap.go:50]: if `Enabled()` can use `zpl.logger.Check()` we could save the string formatting cost by testing before calling `print`

File: zap/zap.go [hunk 3]
```diff
@@ -14,89 +15,173 @@ var (
...
+	zpl.logger = zpl.logger.With(zs...)
+	return zpl
+}
+
+// New creates a slog.Logger adaptor using a zap as backend. If
+// none was passed it will create an opiniated new one, and you will
+// have no option to change its config
+func New(logger *zap.Logger) slog.Logger {
+
+	return newLogger(logger, "info")
```
Comment (amery) [inline zap/zap.go:142]: wouldn't it make more sense to let the user choose the level? maybe using our constants?

File: zap/zap.go [hunk 4]
```diff
@@ -14,89 +15,173 @@ var (
...
+func (zpl *Logger) WithField(label string, value any) slog.Logger {
+	zpl.logger = zpl.logger.With(zap.Any(label, value))
+	return zpl
 }
 
 // WithFields returns a new logger with a set of fields attached
-func (zl *Logger) WithFields(fields map[string]any) slog.Logger {
-	return zl
+func (zpl *Logger) WithFields(fields map[string]any) slog.Logger {
+	zs := make([]zap.Field, 0)
```
Comment (amery) [inline zap/zap.go:129]: we could use len(fields) to preallocate space

File: zap/zap.go [hunk 5]
```diff
@@ -14,89 +15,173 @@ var (
...
+	}
+
+	return zpl
 }
 
 // WithStack attaches a call stack to a new logger
-func (zl *Logger) WithStack(skip int) slog.Logger {
-	return zl
+func (zpl *Logger) WithStack(skip int) slog.Logger {
+	zpl.logger = zpl.logger.WithOptions(zap.AddStacktrace(zapcore.InfoLevel))
```
Comment (amery) [inline zap/zap.go:117]: why `Info` instead of the level chosen by `WithLevel()`?

**Review by amery — CHANGES_REQUESTED** [review]
we also need a LICENCE file here to allow pkg.go.dev to show the module, in which case it would help to merge https://github.com/darvaza-proxy/slog/pull/20 first

File: zap/zap.go (hunk 1 above)
Comment (karasz) [reply zap/zap.go:30]: The "core" has a minimum level, that is what we check here

File: zap/zap.go (hunk 5 above)
Comment (karasz) [reply zap/zap.go:123]: WithStack will enable stack trace for all levels, so I chose the absolute minimal level

File: zap/zap.go (hunk 4 above)
Comment (karasz) [reply zap/zap.go]: done

File: zap/zap.go (hunk 2 above)
Comment (karasz) [reply zap/zap.go]: Optimized to use Enabled()

File: zap/zap.go (hunk 1 above)
Comment (amery) [reply zap/zap.go:30]: Enabled() isn't only avoid being _valid_. it's intended to check if the level specified via `WithLevel()` will be processed and it's worth the processing effort

File: zap/zap.go [hunk 6]
```diff
@@ -14,89 +15,178 @@ var (
...
 }
+
+func newLogger(logger *zap.Logger, lv zap.AtomicLevel) *Logger {
+	var cfg *zap.Config
+
+	if logger == nil {
+		*cfg = zap.NewProductionConfig()
+		cfg.Encoding = "console"
+		cfg.Sampling = &zap.SamplingConfig{}
+		cfg.Level = lv
```
Comment (amery) [inline zap/zap.go:179]: `lv` not relevant when a parent logger is provided?

**Review by amery — APPROVED** [review]

File: zap/zap.go [hunk 7]
```diff
@@ -14,89 +15,176 @@ var (
...
+	var levels = []zapcore.Level{
+		slog.UndefinedLevel: zapcore.InvalidLevel,
+		slog.Fatal:          zapcore.FatalLevel,
+		slog.Error:          zapcore.ErrorLevel,
+		slog.Warn:           zapcore.WarnLevel,
+		slog.Info:           zapcore.InfoLevel,
+		slog.Debug:          zapcore.DebugLevel,
+	}
+
+	if level < -1 || int(level) >= len(levels) {
```
Comment (amery) [inline zap/zap.go:104]: here `level` is a `slog.Level` so `-1` isn't valid.

File: zap/zap.go [hunk 8]
```diff
@@ -14,89 +15,176 @@ var (
...
+		// fix your code
+		err := fmt.Errorf("slog: invalid log level %v", level)
+		zpl.logger.Panic(err.Error())
```
Comment (amery) [inline zap/zap.go:107]: if you are passing a _string_ to `Panic`, why bother creating an `error`?

File: zap/zap.go [hunk 9]
```diff
@@ -14,89 +15,176 @@ var (
...
+func (zpl *Logger) WithStack(skip int) slog.Logger {
+	zpl.logger = zpl.logger.WithOptions(zap.AddStacktrace(zpl.logger.Level()), zap.AddCallerSkip(skip))
+	return zpl
 }
 
 // WithField returns a new logger with a field attached
-func (zl *Logger) WithField(label string, value any) slog.Logger {
-	return zl
+func (zpl *Logger) WithField(label string, value any) slog.Logger {
+	zpl.logger = zpl.logger.With(zap.Any(label, value))
```
Comment (amery) [inline zap/zap.go:124]: Adding fields even if not Enabled()?

File: zap/zap.go [hunk 10]
```diff
@@ -14,89 +15,176 @@ var (
...
 
 // WithFields returns a new logger with a set of fields attached
-func (zl *Logger) WithFields(fields map[string]any) slog.Logger {
-	return zl
+func (zpl *Logger) WithFields(fields map[string]any) slog.Logger {
+	zs := make([]zap.Field, len(fields))
+	for k, v := range fields {
+		zs = append(zs, zap.Any(k, v))
+	}
+	zpl.logger = zpl.logger.With(zs...)
```
Comment (amery) [inline zap/zap.go:134]: Adding fields even if not Enabled()?

File: zap/zap.go [hunk 11]
```diff
@@ -14,89 +15,176 @@ var (
...
+// New creates a slog.Logger adaptor using a zap as backend. If
+// none was passed it will create an opiniated new one.
+func New(logger *zap.Logger, cfg *zap.Config) slog.Logger {
+	return newLogger(logger, cfg)
+}
+
+// NewWithCallback creates a new zap logger using a callback to modify it.
+func (zpl *Logger) NewWithCallback(fn func(lv zapcore.Entry) error) *Logger {
+	if fn != nil {
+		zpl.logger = zpl.logger.WithOptions(zap.Hooks(fn))
```
Comment (amery) [inline zap/zap.go:147]: this would crash if `zpl.logger == nil`

File: zap/zap.go [hunk 12]
```diff
@@ -14,89 +15,176 @@ var (
...
+	}
+
+	return zpl
 }
 
 // WithStack attaches a call stack to a new logger
-func (zl *Logger) WithStack(skip int) slog.Logger {
-	return zl
+func (zpl *Logger) WithStack(skip int) slog.Logger {
+	zpl.logger = zpl.logger.WithOptions(zap.AddStacktrace(zpl.logger.Level()), zap.AddCallerSkip(skip))
```
Comment (amery) [inline zap/zap.go:118]: `skip + 1`, to reflect the extra depth

**Review by amery — CHANGES_REQUESTED** [review]

File: zap/zap.go (hunk 7 above)
Comment (karasz) [reply zap/zap.go:104]: fixed

File: zap/zap.go (hunk 12 above)
Comment (karasz) [reply zap/zap.go:118]: fixed

File: zap/zap.go (hunk 8 above)
Comment (karasz) [reply zap/zap.go]: fixed

File: zap/zap.go (hunk 9 above)
Comment (karasz) [reply zap/zap.go]: fixed

File: zap/zap.go (hunk 10 above)
Comment (karasz) [reply zap/zap.go]: fixed

File: zap/zap.go (hunk 11 above)
Comment (karasz) [reply zap/zap.go:150]: now it is safe

**Review by amery — APPROVED** [review]

**Pull request merged**